### PR TITLE
Update pu/pu to support new Deployment Settings endpoints

### PR DESCRIPTION
**Pull request by glena**
# Description

Add DS endpoints to the backend and client

Fixes https://github.com/pulumi/pulumi-service/issues/20276

## Checklist

- [ ] I have run `make tidy` to update any new dependencies
- [ ] I have run `make lint` to verify my code passes the lint check
  - [ ] I have formatted my code using `gofumpt`

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

**Comment by glena** [comment]
Depends on https://github.com/pulumi/pulumi/pull/16381
I am pushing smaller PRs to make it easier to review

File: pkg/backend/httpstate/client/client.go [hunk 1]
```diff
@@ -1131,6 +1131,37 @@ func (pc *Client) UpdateStackTags(
...
+func (pc *Client) UpdateStackDeploymentSettings(ctx context.Context, stack StackIdentifier,
+	deployment apitype.DeploymentSettings,
+) error {
+	return pc.restCall(ctx, "PUT", getStackPath(stack, "deployments", "settings"), nil, deployment, nil)
+}
+
+func (pc *Client) EncryptStackDeploymentSettingsSecret(ctx context.Context,
+	stack StackIdentifier, secret string,
+) (string, error) {
+	request := apitype.SecretValue{}
```
Comment (julienp) [inline pkg/backend/httpstate/client/client.go:1143]: Should this take the `secret` as Value?

File: pkg/backend/mock.go [hunk 2]
```diff
@@ -374,6 +386,38 @@ func (be *MockBackend) CancelCurrentUpdate(ctx context.Context, stackRef StackRe
 	panic("not implemented")
 }
 
+func (be *MockBackend) EncryptStackDeploymentSettingsSecret(ctx context.Context, stack Stack, secret string) (string, error) {
+	if be.UpdateStackTagsF != nil {
```
Comment (julienp) [inline pkg/backend/mock.go:390]:
```suggestion
	if be.EncryptStackDeploymentSettingsSecretF != nil {
```

File: pkg/backend/mock.go [hunk 3]
```diff
@@ -374,6 +386,38 @@ func (be *MockBackend) CancelCurrentUpdate(ctx context.Context, stackRef StackRe
...
+		return be.EncryptStackDeploymentSettingsSecretF(ctx, stack, secret)
+	}
+	panic("not implemented")
+}
+
+func (be *MockBackend) UpdateStackDeploymentSettings(ctx context.Context, stack Stack,
+	deployment apitype.DeploymentSettings,
+) error {
+	if be.UpdateStackTagsF != nil {
```
Comment (julienp) [inline pkg/backend/mock.go:399]:
```suggestion
	if be.UpdateStackDeploymentSettingsF != nil {
```

File: pkg/backend/mock.go [hunk 4]
```diff
@@ -374,6 +386,38 @@ func (be *MockBackend) CancelCurrentUpdate(ctx context.Context, stackRef StackRe
...
+		return be.UpdateStackDeploymentSettingsF(ctx, stack, deployment)
+	}
+	panic("not implemented")
+}
+
+func (be *MockBackend) GetStackDeploymentSettings(ctx context.Context,
+	stack Stack,
+) (*apitype.DeploymentSettings, error) {
+	if be.UpdateStackTagsF != nil {
```
Comment (julienp) [inline pkg/backend/mock.go:408]:
```suggestion
	if be.GetStackDeploymentSettingsF != nil {
```

File: pkg/backend/mock.go [hunk 5]
```diff
@@ -374,6 +386,38 @@ func (be *MockBackend) CancelCurrentUpdate(ctx context.Context, stackRef StackRe
...
+		return be.GetStackDeploymentSettingsF(ctx, stack)
+	}
+	panic("not implemented")
+}
+
+func (be *MockBackend) DestroyStackDeploymentSettings(ctx context.Context, stack Stack) error {
+	if be.UpdateStackTagsF != nil {
```
Comment (julienp) [inline pkg/backend/mock.go:415]:
```suggestion
	if be.DestroyStackDeploymentSettingsF != nil {
```

**Comment by tgummerer** [comment]
I'm assuming DS stands for deployments looking at the PR? I updated the title accordingly to make it a bit easier to follow, but please update it if that's incorrect.

File: pkg/backend/diy/backend.go [hunk 6]
```diff
@@ -1350,6 +1354,32 @@ func (b *diyBackend) UpdateStackTags(ctx context.Context,
...
+func (b *diyBackend) UpdateStackDeploymentSettings(ctx context.Context, stack backend.Stack,
+	deployment apitype.DeploymentSettings,
+) error {
+	// The local backend does not support managing deployments.
+	return errors.New("stack deployments not supported with diy backends")
+}
+
+func (b *diyBackend) DestroyStackDeploymentSettings(ctx context.Context, stack backend.Stack) error {
+	// The local backend does not support managing deployments.
+	return errors.New("stack deployments not supported in --local mode")
```
Comment (tgummerer) [inline pkg/backend/diy/backend.go:1373]:
```suggestion
	return errors.New("stack deployments not supported with diy backends")
```

File: pkg/backend/diy/backend.go [hunk 7]
```diff
@@ -1350,6 +1354,32 @@ func (b *diyBackend) UpdateStackTags(ctx context.Context,
...
+}
+
+func (b *diyBackend) GetStackDeploymentSettings(ctx context.Context,
+	stack backend.Stack,
+) (*apitype.DeploymentSettings, error) {
+	// The local backend does not support managing deployments.
+	return nil, errors.New("stack deployments not supported in --local mode")
```
Comment (tgummerer) [inline pkg/backend/diy/backend.go:1380]:
```suggestion
	return nil, errors.New("stack deployments not supported with diy backends")
```

**Review by tgummerer — APPROVED** [review]
Couple of error messages that I think need updating, other than that it looks good to me.

File: pkg/backend/diy/backend.go (hunk 7 above)
Comment (glena) [reply pkg/backend/diy/backend.go:1380]: I was sure I updated those initially and turns out I missed two 🤦

**Comment by glena** [comment]
DS stands for deployment settings, but in this context is something new :) I updated the title

**Pull request merged**